File: src/plugins/registry.rs
use indexmap::IndexMap;
use serde::{Serialize, Deserialize};

pub type InstrId = u32;

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
pub enum Scope { Thread, Subgroup, Workgroup }
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
pub enum SyncLevel { Independent, Synchronized }
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
pub enum MemSem { None, Acquire, Release, AcqRel }

#[derive(Debug, Clone)]
pub struct OperandDesc { pub name: String, pub ty: String }
#[derive(Debug, Clone)]
pub struct ResultDesc  { pub name: String, pub ty: String }

#[derive(Debug, Clone)]
pub struct InstructionSpec {
    pub name: String,
    pub operands: Vec<OperandDesc>,
    pub results: Vec<ResultDesc>,
    pub has_scope: bool,
    pub has_sync: bool,
    pub has_memsem: bool,
    pub needs_subgroup: bool,
    pub needs_shared_mem: bool,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct Params {
    pub scope: Option<Scope>,
    pub sync: Option<SyncLevel>,
    pub mem: Option<MemSem>,
    #[serde(flatten)]
    pub rest: IndexMap<String, serde_json::Value>,
}

pub trait InterpreterHandler: Send + Sync {
    fn interpret(&self, id: InstrId, params: &Params,
                 inputs: &[serde_json::Value],
                 outputs: &mut Vec<serde_json::Value>,
                 ctx: &mut crate::semantics::SemanticsContext) -> bool;
    fn expand_to_core(&self, _id: InstrId, _params: &Params) -> Option<()> { None }
}

pub struct Registry {
    specs: IndexMap<InstrId, InstructionSpec>,
    ids: IndexMap<String, InstrId>,
    next: InstrId,
    handlers: IndexMap<String, IndexMap<InstrId, Box<dyn InterpreterHandler>>>, // model -> (id->handler)
}

impl Registry {
    pub fn new() -> Self { Self { specs: IndexMap::new(), ids: IndexMap::new(), next: 1, handlers: IndexMap::new() } }
    pub fn register_instruction(&mut self, spec: InstructionSpec) -> InstrId {
        let id = self.next; self.next += 1;
        self.ids.insert(spec.name.clone(), id);
        self.specs.insert(id, spec); id
    }
    pub fn register_handler(&mut self, model: &str, id: InstrId, h: Box<dyn InterpreterHandler>) {
        self.handlers.entry(model.to_string()).or_default().insert(id, h);
    }
    pub fn lookup(&self, name: &str) -> Option<InstrId> { self.ids.get(name).copied() }
    pub fn handler(&self, model: &str, id: InstrId) -> Option<&(dyn InterpreterHandler)> {
        self.handlers.get(model)?.get(&id).map(|b| &**b)
    }
}

File: lib/plugins/Registry.cpp
#include "simt-step/plugins/Registry.h"

#include "simt-step/semantics/SemanticsContext.h"

#include <llvm/Support/Error.h>

namespace simt::plugins {

using namespace llvm;

Registry::Registry() = default;

InstructionId Registry::registerInstruction(InstructionSpec spec) {
    const InstructionId id = nextId_++;
    ids_.insert_or_assign(spec.name, id);
    specs_.insert_or_assign(id, std::move(spec));
    return id;
}

void Registry::registerHandler(StringRef model,
                               InstructionId id,
                               std::shared_ptr<const InterpreterHandler> handler) {
    handlers_[model].insert_or_assign(id, std::move(handler));
}

std::optional<InstructionId> Registry::lookup(StringRef name) const {
    if (auto it = ids_.find(name); it != ids_.end()) {
        return it->second;
    }
    return std::nullopt;
}

const InterpreterHandler *Registry::getHandler(StringRef model, InstructionId id) const {
    if (auto modelIt = handlers_.find(model); modelIt != handlers_.end()) {
        if (auto handlerIt = modelIt->second.find(id); handlerIt != modelIt->second.end()) {
            return handlerIt->second.get();
        }
    }
    return nullptr;
}

static Expected<core::Scope> parseScope(StringRef value) {
    if (value.equals_insensitive("thread")) {
        return core::Scope::Thread;
    }
    if (value.equals_insensitive("subgroup")) {
        return core::Scope::Subgroup;
    }
    if (value.equals_insensitive("workgroup")) {
        return core::Scope::Workgroup;
    }
    return make_error<StringError>("unknown scope value", inconvertibleErrorCode());
}

static Expected<SyncLevel> parseSyncLevel(StringRef value) {
    if (value.equals_insensitive("independent")) {
        return SyncLevel::Independent;
    }
    if (value.equals_insensitive("synchronized")) {
        return SyncLevel::Synchronized;
    }
    return make_error<StringError>("unknown sync level", inconvertibleErrorCode());
}

static Expected<core::MemorySemantics> parseMemSem(StringRef value) {
    if (value.equals_insensitive("none")) {
        return core::MemorySemantics::None;
    }
    if (value.equals_insensitive("acquire")) {
        return core::MemorySemantics::Acquire;
    }
    if (value.equals_insensitive("release")) {
        return core::MemorySemantics::Release;
    }
    if (value.equals_insensitive("acqrel") || value.equals_insensitive("acquire_release")) {
        return core::MemorySemantics::AcquireRelease;
    }
    return make_error<StringError>("unknown memory semantics", inconvertibleErrorCode());
}

Expected<Params> parseParams(const json::Value &value) {
    const auto *object = value.getAsObject();
    if (!object) {
        return make_error<StringError>("expected params object", inconvertibleErrorCode());
    }

    Params params;
    for (const auto &entry : *object) {
        const StringRef key = entry.first;
        if (key == "scope") {
            if (auto str = entry.second.getAsString()) {
                auto parsed = parseScope(*str);
                if (!parsed) {
                    return parsed.takeError();
                }
                params.scope = *parsed;
            }
            continue;
        }
        if (key == "sync") {
            if (auto str = entry.second.getAsString()) {
                auto parsed = parseSyncLevel(*str);
                if (!parsed) {
                    return parsed.takeError();
                }
                params.sync = *parsed;
            }
            continue;
        }
        if (key == "mem") {
            if (auto str = entry.second.getAsString()) {
                auto parsed = parseMemSem(*str);
                if (!parsed) {
                    return parsed.takeError();
                }
                params.memory = *parsed;
            }
            continue;
        }

        params.rest.try_emplace(key, entry.second);
    }

    return params;
}

} // namespace simt::plugins
